stdlib/immutablemap.gr

```diff
@@ -0,0 +1,484 @@
+/**
+ * @module ImmutableMap: An ImmutableMap holds key-value pairs. Any value may be used as a key or value. Operations on an ImmutableMap do not mutate the map's internal state.
+ * @example import ImmutableMap from "immutablemap"
+ *
+ * @since v0.5.4
+ */
+
+import List from "list"
+import Array from "array"
+
+// implementation based on the paper "Implementing Sets Efficiently in a
+// Functional Language" by Stephen Adams
+
+record Node<k, v> {
+  key: k,
+  val: v,
+  size: Number,
+  left: ImmutableMap<k, v>,
+  right: ImmutableMap<k, v>,
+},
+enum ImmutableMap<k, v> {
+  Empty,
+  Tree(Node<k, v>),
+}
+
+/**
+ * @section Values: Functions for working with ImmutableMaps.
+ */
+
+// semi-arbitrary value chosen for algorithm for determining when to balance
+// trees; no tree can have a left subtree containing this number of times
+// more elements than its right subtree or vice versa
+let weight = 4
+
+/**
+ * Creates a new, empty map.
+ *
+ * @returns An empty map
+ *
+ * @since v0.5.4
+ */
+export let make = () => {
+  Empty
+}
+
+// returns the key-value pair of the minimum key in a tree
+let rec min = node => {
+  match (node) {
+    Tree({ key, val, left: Empty, _ }) => (key, val),
+    Tree({ left, _ }) => min(left),
+    Empty => fail "Impossible: min of empty element in ImmutableMap",
+  }
+}
+
+/**
+ * Provides the count of key-value pairs stored within the map.
+ *
+ * @param map: The map to inspect
+ * @returns The count of key-value pairs in the map
+ *
+ * @since v0.5.4
+ */
+export let size = map => {
+  match (map) {
+    Empty => 0,
+    Tree({ size, _ }) => size,
+  }
+}
+
+/**
+ * Determines if the map contains no key-value pairs.
+ *
+ * @param map: The map to inspect
+ * @returns `true` if the given map is empty or `false` otherwise
+ *
+ * @since v0.5.4
+ */
+export let isEmpty = map => {
+  size(map) == 0
+}
+
+// helper function for creating a tree node with correct size from
+// two balanced trees
+let makeNode = (key, val, left, right) => {
+  Tree({ key, val, size: 1 + size(left) + size(right), left, right })
+}
+
+// note: see Figure 1 of paper referenced above for visual illustration of
+// the below rotations
+
+// node rotation moving the left subtree of the right node to the left side
+let singleL = (key, val, left, right) => {
+  let Tree({ key: rKey, val: rVal, left: rl, right: rr, _ }) = right
```

Comment:
I believe this causes compiler warnings since it's not exhaustive. It's a bit safer to do a pattern match for assignments like these.

Reply:
In this case it _should_ be impossible for the tree to be empty; perhaps I'll mimic an `Option.expect` here with a custom "Impossible: ..." message so that it's at least easier to tell what's going wrong if there's a bug?

Reply:
Also as a follow-up this case actually doesn't cause a compiler warning